dll/win32/iernonce/registry.cpp

```diff
@@ -0,0 +1,206 @@
+/*
+ * PROJECT:     ReactOS system libraries
+ * LICENSE:     GPL - See COPYING in the top level directory
+ * FILE:        dll\win32\iernonce\registry.cpp
+ * PURPOSE:     Functions to read RunOnceEx registry.
+ * PROGRAMMERS: Copyright 2013-2016 Robert Naumann
+ */
+
+
+#include <cassert>
+#include "registry.h"
+#include <winreg.h>
+
+
+typedef BOOL(*REGENUMCALLBACK)(HKEY hKey, LPCWSTR lpObjectName, DWORD cchszName, PVOID Param);
+
+enum class REG_ENUM_TYPE
+{
+    SUBKEY,
+    VALUE
+};
+
+
+bool RunOnceExSectionCmp::operator()(const pRunOnceExSection &lhs, const pRunOnceExSection &rhs) const
+{
+    return lstrcmpW(lhs->m_SectionName.GetString(), rhs->m_SectionName.GetString()) < 0;
+}
+
+bool RunOnceExEntryCmp::operator()(const ATL::CStringW &lhs, const ATL::CStringW &rhs) const
+{
+    return lstrcmpW(lhs.GetString(), rhs.GetString()) < 0;
+}
+
+BOOL RegEnumObject(HKEY hKey, REG_ENUM_TYPE EnumType, REGENUMCALLBACK Callback, PVOID Param)
+{
+    DWORD dwObjectNum;
+    DWORD dwMaxObjNameLen;
+    LSTATUS Error;
+    switch (EnumType)
+    {
+        case REG_ENUM_TYPE::SUBKEY:
+            Error = RegQueryInfoKeyW(hKey, NULL, 0, NULL,
+                &dwObjectNum, &dwMaxObjNameLen, NULL, NULL, NULL, NULL, NULL, NULL);
+            break;
+        case REG_ENUM_TYPE::VALUE:
+            Error = RegQueryInfoKeyW(hKey, NULL, 0, NULL,
+                NULL, NULL, NULL, &dwObjectNum, &dwMaxObjNameLen, NULL, NULL, NULL);
+            break;
+        default:
+            assert(0);
+            return FALSE;
+    }
+    if (Error != ERROR_SUCCESS)
+    {
+        return FALSE;
+    }
+
+    LPWSTR RegObjectName = (LPWSTR)HeapAlloc(GetProcessHeap(),
```

Comment:
Is the cast needed?

Reply:
yes, I guess. If it's in C this is not necessary but C++ seems to be more strict 🤣 (at least my compiler won't let go of it)

Reply:
This is c++, and you are using ATL already, so I would suggest to use CComHeapPtr or the likes instead of writing code in all exit paths to delete this.

Reply:
Yes indeed, now I just use a simple `CStringW` instead 😄